### PR TITLE
feature(messaging): android messaging upgrade

### LINKED ISSUE #3209: 🔥[Android only] Notifications appear in notification bar but not in lock screen when phone is locked
Hi!

This is only happening on Android. 
I get notification from other apps like WhatsApp or Instagram when my phone is locked and I also hear a sound when they arrive. However, when I schedule a notification and the phone is locked **I won't receive either a notification on the lock screen nor will I hear a sound.** 
However, once you unlock the phone you can see the notification in the top notification bar and expand it.

Is this a bug?

This is the code to create the notification:
````
 // Builds and create notification channel for android devices
    if (Platform.OS === 'android') {
      // Build the channel
      const channel = new firebase.notifications.Android.Channel('moodPixel-channel', 'Daily Reminder', firebase.notifications.Android.Importance.High)
      .setDescription('Daily notification reminders').setSound('default');
      // Create the channel
      firebase.notifications().android.createChannel(channel);
    }
  
    // Set recurrent reminder notification
    const notification = new firebase.notifications.Notification({
      sound: 'default',
      show_in_foreground: true,
    })
    .setNotificationId('reminder-notification')
    .setTitle(strings.Notifications.mainNotificationTitle)
    .setBody(strings.Notifications.message)
    .setSound('default');
  
    // add channelId if on android
    if (Platform.OS === 'android') {
      notification.android.setChannelId('moodPixel-channel')
      notification.android.setPriority(firebase.notifications.Android.Priority.High);
      notification.android.setSmallIcon('@mipmap/ic_notif');
    }
  
    date.setDate(date.getDate());
    date.setHours(notifHour); 
    date.setMinutes(notifMinute);
    date.setMilliseconds(0);
  
    firebase.notifications().scheduleNotification(notification, {
      fireDate: date.getTime(),
      repeatInterval: 'day',
    });
````

npm
````
"react-native": "0.60.4",
"react-native-firebase": "^5.5.6",
````

Any ideas how to make this work?

### PR DESCRIPTION
**Pull request by Ehesp**
This upgrade includes the following changes:

- Rename of `registerForRemoteNotifications` to `registerDeviceForRemoteMessages`. This will be a deprecation and will be removed in v7. Nothing internally is changing, however the current wording is just totally confusing.
- The background handler will be ALWAYS set, and if you dont provide a handler to `setBackgroundMessageHandler` it'll throw a better/custom warning (e.g. not a React Native one).
- `onMessage` will work for data-only, notification-only AND data-notification messages. Currently it only works for data-only.
- `setBackgroundMessageHandler` will now work when the app is both fully quit and backgrounded, for data-only, notification-only AND data-notification messages. 
- We're implementing both `getInitialNotification` & `onNotificationOpenedApp` into the messaging module. This won't require any native changes from users.
- If a message is sent with a notification payload, it'll be included within the `RemoteMessage` object passed around.

Fixes #3209, #3032, #3293, #3236

### CODE REVIEW

**Comment by mikehardy** [comment]
This sounds amazing! If only github made it easy to do a plot of "new issues over time" so you could watch the change in that rate before this lands and after. Huge

**Comment by Ehesp** [comment]
I'm also updating the new docs to reflect the changes: https://rnfb-docs.netlify.com/messaging/notifications

Not sure when that'll go live, but if it's not until v7 I'll extract aspects into the current docs for v6.

File: packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingHeadlessService.java [hunk 1]
```diff
@@ -25,7 +25,9 @@ HeadlessJsTaskConfig getTaskConfig(Intent intent) {
       TASK_KEY,
       ReactNativeFirebaseMessagingSerializer.remoteMessageToWritableMap(remoteMessage),
       ReactNativeFirebaseJSON.getSharedInstance().getLongValue(TIMEOUT_JSON_KEY, TIMEOUT_DEFAULT),
-      false
+      // Prevents race condition where the user opens the app at the same time as a notification
+      // is delivered, causing a crash.
+      true
```
Comment (Ehesp) [inline packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingHeadlessService.java]: https://github.com/invertase/react-native-firebase/pull/3311

**Comment by Stringsaeed** [comment]
@Ehesp when installing **6.4.0-rc0** get build error 
`> Task :app:processDebugManifest
/Users/stringsaeed/WebstormProjects/help/android/app/src/debug/AndroidManifest.xml:24:9-31:50 Warning:
	activity#com.google.firebase.auth.internal.FederatedSignInActivity@android:launchMode was tagged at AndroidManifest.xml:24 to replace other declarations but no other declaration present
`
```FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processDebugResources'.
> Android resource linking failed
  /Users/stringsaeed/WebstormProjects/help/android/app/build/intermediates/merged_manifests/debug/AndroidManifest.xml:74: AAPT: error: '' is incompatible with attribute resource (attr) reference.
      
  error: failed processing manifest.

**Comment by mikehardy** [comment]
@Stringsaeed the error looks like it is local to your project, the warning related to firebase would not cause a failure

Recommend `cd android && ./gradlew clean` and retry, if it continues failing then somewhere in your AndroidManifest you are referencing a string with an empty reference I think? Android Studio might help with making it more obvious if you open the manifest there

**Comment by Stringsaeed** [comment]
@mikehardy it's related messaging_android_notification_channel_id and messaging_android_notification_color in `firebase.json` it should be required

**Comment by Stringsaeed** [comment]
@Ehesp so `onNotificationOpenedApp()` just work when in background but if i closed the app it wouldn't work

**Comment by Ehesp** [comment]
@Stringsaeed did you get it working with no channel/color?

Also when you say "background", do you mean app is still active but not visible?

**Comment by Ehesp** [comment]
>  if i closed the app it wouldn't work

There us no bridge to send an event, you must call `getInitialNotification`

**Comment by Stringsaeed** [comment]
@Ehesp 
no it didn't worked with no channel or color i added channel and color to make it work
yes background means app is still active but not visible notification opend app working like a charm but if app is inactive at all not launched so notification opend app not working

also onMessage fires when onNotificationOpenedApp fires  when only app is in background

**Comment by Ehesp** [comment]
Ok thanks, what android device are you building on? I'll dig into that issue with color/id.

When the app is quit, you cannot get events via `onMessage` or `onNotificationOpenedApp`. When the user opens the app via the notification when it's quit, you need to call `getInitialNotification`.

**Comment by Stringsaeed** [comment]
Samsung Galaxy A70, Android version 9,
thank you for your response

**Comment by carltran** [comment]
@Stringsaeed Hi, I having exactly the same error 'AAPT...' you encountered above. Can you pls provide details how you resolved it? Thanks

**Comment by Stringsaeed** [comment]
@carltran it's in new documentation you should add  [`firebase.json` check the link here](https://github.com/invertase/react-native-firebase/blob/master/tests/firebase.json) in your project root and add this 
```json
{
  "react-native": {
    "messaging_auto_init_enabled": true,
    "messaging_android_headless_task_timeout": 30000,
    "messaging_android_notification_channel_id": "",
    "messaging_android_notification_color": "@color/white",
  }
}
```
you should add the name of your color in android/app/src/main/res/values/colors.xml

like if you specified "messaging_android_notification_color": "@color/white"
your colors.xml should be looks like this 
```xml
<?xml version="1.0" encoding="utf-8"?>
<resources>
   ...
    <color name="white">#FFFFFF</color>
   ...
</resources>
```

**Comment by carltran** [comment]
@Stringsaeed it works. thanks a lot.

@Ehesp thanks for the great work implementing this background notification handling. Any idea about the timeline of release 6.4?

**Comment by Ehesp** [comment]
iOS still needs implementing which is taking a bit of time. Also these issues which are being highlighted need sorting. The more who use it, the faster it'll come.

**Comment by Stringsaeed** [comment]
@Ehesp if you need anything I can help to speed up the process.

**Comment by Ehesp** [comment]
@Stringsaeed thanks - PR #3339 up to solve that problem. For now, just enter `@color/white`. We also provide default colors so you dont need to create a resource file unless you need something custom.

**Comment by pierregoutheraud** [comment]
@Stringsaeed I am getting an error related to my firebase.json config
```
{
  "react-native": {
    "messaging_android_notification_channel_id": "default",
    "messaging_android_notification_color": "@color/main"
  }
}
```

It is working on debug but on Android release build I am getting this error :
```
Android resource linking failed
/Users/pierregoutheraud/www/alike-app/node_modules/@react-native-firebase/messaging/android/build/intermediates/aapt_friendly_merged_manifests/release/aapt/AndroidManifest.xml:40:
error: resource color/main (aka io.invertase.firebase.messaging:color/main) not found.
```

I have defined `@color/main` in my `colors.xml`
```
<?xml version="1.0" encoding="utf-8"?>
<resources>
    <color name="main">#6147FF</color>
</resources>
```
 
Looks like it's not looking for color/main at the right place ?

**Comment by Ehesp** [comment]
Can you check the latest pr, there's been a few fixes, including one for this.

**Comment by pierregoutheraud** [comment]
@Ehesp Thank you

**Comment by pierregoutheraud** [comment]
You mean https://github.com/invertase/react-native-firebase/pull/3339 which is on `6.4.0-rc2` ?
Because I am still getting the error on `6.4.0-rc3`

**Comment by Ehesp** [comment]
If you open the android directory on android studio, then open the manifest file. If the resource can't be found, it'll tell you.

**Comment by pierregoutheraud** [comment]
@Ehesp I am sorry, I am not sure I understand.
If you are talking about my main AndroidManifest file, then I do not use the `@color/main` resource there.
Can I use anything other than `@color/white` in `firebase.json` on `6.4.0-rc3` ?

**Comment by vieiralucas** [comment]
@pierregoutheraud did you ever managed to fix this?
I'm on `v7.1.0` and experiencing the exact same problem

**Comment by alexbuicescu** [comment]
I get the same errors on **7.1.0**, with react-native: **0.61.5**.

`gradlew installDebug` works successfully, but `gradlew assembleRelease` throws:
```
* What went wrong:
Execution failed for task ':react-native-firebase_messaging:verifyReleaseResources'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
   > Android resource linking failed
     /Users/alexbuicescu/project/node_modules/@react-native-firebase/messaging/android/build/intermediates/aapt_friendly_merged_manifests/release/aapt/AndroidManifest.xml:37: AAPT: error: resource string/default_notification_channel_id (aka io.invertase.firebase.messaging:string/default_notification_channel_id) not found.
         
     /Users/alexbuicescu/project/node_modules/@react-native-firebase/messaging/android/build/intermediates/aapt_friendly_merged_manifests/release/aapt/AndroidManifest.xml:40: AAPT: error: resource color/colorAccent (aka io.invertase.firebase.messaging:color/colorAccent) not found.
```

And here are my resources, you can see **colorAccent** and **default_notification_channel_id** definitely exist (android/app/src/main/res/values/):
1. colors.xml:
```xml
<?xml version="1.0" encoding="utf-8"?>
<resources>
  <color name="colorPrimary">#ff0000</color>
  <color name="colorPrimaryDark">#ff0000</color>
  <color name="colorAccent">#ff0000</color>
</resources>
```
2. strings.xml:
```xml
<resources>

  <string name="default_notification_channel_id">AppName</string>

  <string name="app_name">AppName</string>
  <string name="dev_activity_name">Exponent Dev</string>
  <string name="preference_file_key">host.exp.exponent.SharedPreferences</string>
  <string name="error_header">Something went wrong.</string>
  <string name="error_default_client">Sorry about that. You can go back to Expo home or try to reload the Experience.</string>
  <string name="error_default_shell">Sorry about that. Press the reload button to try again.</string>
  <string name="error_unable_to_load_experience">Unable to load Experience.</string>
  <string name="error_uncaught">Uncaught Error: %s</string>
  <string name="view_error_log">View error log</string>
  <string name="default_notification_channel_group">Default</string>
  <string name="persistent_notification_channel_name">Experience notifications</string>
  <string name="persistent_notification_channel_desc">Persistent notifications that provide debug info about open experiences</string>

</resources>
```

And the firebase.json file:
```json
{
  "react-native": {
    "messaging_android_headless_task_timeout": 60000,
    "messaging_android_notification_channel_id": "@string/default_notification_channel_id",
    "messaging_android_notification_color": "@color/colorAccent"
  }
}
```

**Comment by mikehardy** [comment]
That reminded me of something. https://twitter.com/fastsquatch/status/1196475279688028160

It smells like a proguard issue if debug works and release doesn't. Resource shrinking in proguard builds is something most don't pay attention to (see above 3 tweets)

**Comment by iscosta** [comment]
I had the same problem when trying to generate the release apk, so I changed the color path and works fine!

from:
android/app/src/main/res/values/color.xml : 
<color name="colorPrimaryDark">#000051</color>

to:

node_modules/@react-native-firebase/messaging/android/src/main/res/values/colors.xml :
add color -> <color name="colorPrimaryDark">#000051</color>

**Comment by devoto13** [comment]
I don't think this is related to proguard (which is disabled in my project), but rather that dependency references an application resource, which is not available during the dependency build. See https://github.com/invertase/react-native-firebase/issues/3559#issuecomment-686716098.

**Pull request merged**